### ISSUE TITLE
Own data

### ISSUE THREAD

**Issue by salbatron** [adambielski/siamese-triplet#12]
Is there a way to train the network with my own data?

**Comment by sabrinazuraimi** [comment]
I'm also trying to run this with my own image folder(which has a huge number of classes) but I'm having problems with the triplet selections. Anyone have any idea how to implement the triplet selection?

**Comment by madarax64** [comment]
Hello @salbatron @sabrinazuraimi ,
I've also been working on this implementation for short while now, but specifically for the Triplet Network not the Siamese one. To train with your own data, I'd say the easiest way to get this working would be to:

a. Implement a custom Dataset class - specifically, inherit from the base Dataset class and implement the __len__ and __getitem__ methods as described in the [docs](https://pytorch.org/docs/stable/data.html)

b. It may be necessary to implement a BalancedBatchSampler class as well, or simply edit the base version from the repo. This is because in the source code, within the constructor of the BalancedBatchSampler, it is apparently expected that the Dataset class/implementation used has a property called `train_labels` or `test_labels`, the contents of which are stored in the `labels` property of the BalancedBatchSampler. Therefore your Dataset implementation will have to include some way of getting the labels of the instances therein, and you'll need to edit the BalancedBatchSampler to populate its labels property using your chosen method/property.

c. When all that;s done, the already-implemented Triplet selections should work out of the box as far as I can tell.

I hope this was a little helpful to either of you guys.

Good luck!

**Comment by vishalgolcha** [comment]
I was thinking of creating a  pull request with a custom dataloader that i made with this repository if that could help some people  .

**Comment by WillDamon** [comment]
@vishalgolcha 
Would you mind share your code with us on how to train this project with our own data, which is just like  train/1/001.jpg, train/1/002.jpg, ..., train/100/001.jpg, test/101/001.jpg, test/101/002.jpg, ..., test/110/001.jpg...
Thank you so much.

**Comment by Ella77** [comment]
@vishalgolcha Thanks for own data custom generator :) I have some questions regarding this and sent email for you. Would you mind opening it?

**Comment by adambielski** [comment]
@madarax64 explained it very well - thank you for this.
I refactored BalancedBatchSampler, now it should be easier to use with different datasets (it just needs labels in the same order as in the dataset.

All in all, what you need is (for the case with online triplet sampling):
1. A Dataset class, so own implementation of `__len__` and `__getitem__(index)` methods; this is PyTorch specific and you can use some default PyTorch loaders or customize them for your case (see post by @madarax64). See some examples of custom Datasets https://github.com/utkuozbulak/pytorch-custom-dataset-examples
2. BalancedBatchSampler that now takes a list of all labels in the dataset, where n-th label corresponds to n-th index in the Dataset class. The inefficient to get the list of labels would be `labels = [dataset[i][1] for i in range(len(dataset))]`, assuming that `__getitem__(index)` returns a pair of image and a label. Most likely your dataset already stores that list and it can be accessed without processing the images.

**Comment by josianerodrigues** [comment]
Hi @vishalgolcha,

I need to do tests with the CIFAR100, but I can't make the function to generate the triplets and pairs of image work. Can you help me? I believe it changes little because this dataset also has in the pytorch library.
